src/adapter/src/coord/sequencer/alter_set_cluster.rs

```diff
@@ -0,0 +1,89 @@
+use mz_sql::catalog::ObjectType;
+use mz_sql::plan::AlterSetClusterPlan;
+use std::iter::once;
+
+use crate::catalog::CatalogItem;
+use crate::coord::Coordinator;
+use crate::session::Session;
+use crate::{catalog, AdapterError, ExecuteResponse, TimestampProvider};
+
+impl Coordinator {
+    /// Convert a [`AlterSetClusterPlan`] to a sequence of catalog operators and adjust state.
+    pub(super) async fn sequence_alter_set_cluster(
+        &mut self,
+        session: &Session,
+        AlterSetClusterPlan { id, set_cluster }: AlterSetClusterPlan,
+    ) -> Result<ExecuteResponse, AdapterError> {
+        let entry = self.catalog().get_entry(&id);
+        match entry.item() {
+            CatalogItem::MaterializedView(catalog::MaterializedView {
+                optimized_expr,
+                cluster_id,
+                ..
+            }) => {
+                if !self
+                    .catalog()
+                    .state()
+                    .is_system_schema_specifier(&entry.name().qualifiers.schema_spec)
+                    && !self.is_compute_cluster(set_cluster)
+                {
+                    let cluster_name = self.catalog().get_cluster(set_cluster).name.clone();
+                    return Err(AdapterError::BadItemInStorageCluster { cluster_name });
+                }
+
+                let cluster_id = *cluster_id;
+
+                // Pick the least valid read timestamp as the as-of for the view
+                // dataflow. This makes the materialized view include the maximum possible
+                // amount of historical detail.
+                let id_bundle = self
+                    .index_oracle(set_cluster)
+                    .sufficient_collections(&optimized_expr.depends_on());
+                let as_of = self.least_valid_read(&id_bundle);
+
+                // Allocate a unique ID that can be used by the dataflow builder to
+                // connect the view dataflow to the storage sink.
+                let internal_view_id = self.allocate_transient_id()?;
+                let op = catalog::Op::AlterSetCluster {
+                    id,
+                    cluster: set_cluster,
+                };
+                // Update catalog
+                let mut df = self
+                    .catalog_transact_with(Some(session), vec![op], |txn| {
+                        // Create a dataflow that materializes the view query and sinks
+                        // it to storage.
+                        txn.dataflow_builder(set_cluster)
+                            .build_materialized_view_dataflow(id, internal_view_id)
+                    })
+                    .await?;
+
+                df.set_as_of(as_of);
+
+                self.migrate_compute_ids_in_timeline(once((cluster_id, id)), set_cluster);
+                self.drop_collections(vec![id], cluster_id)?;
+
+                // Send the dataflow to the new cluster.
+                self.must_ship_existing_dataflows(once(df), set_cluster);
```

Comment:
Creating a new dataflow with the same ID as the just dropped dataflow will lead to problems. We don't have idempotency, so the controller won't just ignore a `create_dataflows` with a previous ID at least. But I think the controller will become confused anyway.

It's important to note that `drop_collections` doesn't immediately drop the associated `CollectionState`. It just sets the read policy to the empty frontier. If something else still has a read hold on the collection, the `Collection State` won't be dropped until that read hold is released. And it's not enough to make sure the adapter doesn't have any read holds on the collection anymore, as the compute controller itself will also install read holds for collections that depend on the MV. It looks like we can only solve this by adding a `change_instance` API to the compute controller, to let it make the necessary internal changes.

It seems that we don't do a good job in `create_dataflows` to check if the created collection still exists. We will just overwrite the existing `CollectionState` with a new one. We should fix this by returning an error instead, so we panic instead of corrupting the controller state.

Reply:
I fixed this by introducing a uuid that associates `CreateDataflow` commands with reported frontiers. I don't think this is fixes the whole problem you're describing, but it avoids the specific problem of confusing the controller in race conditions where we might have outstanding responses while dropping a collection, and moving it back to the original cluster. Let's determine whether this is enough for this PR, or if we need the `change_instance` API now.